### ISSUE TITLE
Breaking Change: Query crashes when compiled with VS2015 RTM (Roslyn compiler)

### ISSUE THREAD

**Issue by rowanmiller** [aspnet/EntityFramework6#20]
The following query crashes when compiled with VS2015 RTM, but works when compiled with the old compiler in VS2013: 

```
public async Task<List<ProductRelation>> FailingQuery(SampleEnum queryOption)
        {
            using (var context = new SampleContext())
            {

                int[] ids = await Task.FromResult(new[] { 1, 2, 3, 4 });
                return await (from order in context.Set<Order>()
                              where ids.Contains(order.OrderId)
                                 && queryOption == SampleEnum.Value1
                              from pr in context.Set<ProductRelation>()
                              select pr).ToListAsync();
            }
        }
```

(Not that the query has been simplified for demonstration purposes, it obviously makes no sense in this shape but I have reduced it as much as possible).

```
Exception: 

System.NotSupportedException: LINQ to Entities does not recognize the method 'System.Data.Entity.DbSet`1[RoslynEfTest.UnitTest1+ProductRelation] Set[ProductRelation]()' method, and this method cannot be translated into a store expression.
```

I have attached a sample unit-test project which illustrates the error.

**Comment by rowanmiller** [comment]
Ported from CodePlex https://entityframework.codeplex.com/workitem/2785

**Comment by jcchalte** [comment]
Hello,

Is there any news about this bug ? We have a huge codebase and cannot migrate to VS2015 with this bug active. I have retrieved the solution provided in the codeplex forum and tried the code given in both VS 2013 and VS 2015. Indeed, the code does not work in VS2015.

It seems that the expression that EF is trying to convert is not the same in both environments :

In VS2015, this is as following :
`value(System.Data.Entity.Core.Objects.ObjectQuery``1[RoslynEfTest.UnitTest1+Order]).MergeAs(AppendOnly).MergeAs(AppendOnly).Where(order => (value(System.Int32[]).Contains(order.OrderId) AndAlso ([-1] == 0))).SelectMany(order => value(RoslynEfTest.UnitTest1+SampleContext).Set(), (order, pr) => pr)
`
In VS2013, the expression is :
`value(System.Data.Entity.Core.Objects.ObjectQuery``1[RoslynEfTest.UnitTest1+Order]).MergeAs(AppendOnly).MergeAs(AppendOnly).Where(order => (value(System.Int32[]).Contains(order.OrderId) AndAlso ([-1] == 0))).SelectMany(order => value(System.Data.Entity.Core.Objects.ObjectQuery``1[RoslynEfTest.UnitTest1+ProductRelation]).MergeAs(AppendOnly), (order, pr) => pr)`


The relevant difference is that in one case, the Set method is interprated as a "MergeAs" call to an objectquery 

`value(System.Data.Entity.Core.Objects.ObjectQuery``1[RoslynEfTest.UnitTest1+ProductRelation]).MergeAs(AppendOnly)`

whereas in the other case, it is still a call to the Set method.

` value(RoslynEfTest.UnitTest1+SampleContext).Set()`

It appears that the expression provided to DbQueryProvider.CreateQuery is not the same in both environments. I still have to figure what called this CreateQuery method.

**Comment by jcchalte** [comment]
Hi again

I have found the root of this issue. In the `DbQueryVisitor` class, you traverse the expression tree in order to replace all the calls to context "Set" methods by the invocation of those methods (in `VisitMethodCall` method). If the visited node is a member expression, you retrieve the the context instance through the `GetContextFromConstantExpression` method. The issue is in this method.

In this method, you assume that the left operand of the expression is a ConstantExpression, otherwise you return null. This was the case in VS2013 as the context instance was directly integrated in a constant. Since VS2015 (probably since Roslyn ?), the context provided in the code above is not directly embedded in a constant but rather in a field of this constant.

A simplified example would be :
in VS 2013, the expression tree was `value(SomeRuntimeTypeVariable).context.Set`, but now it is `value(SomeRuntimeTypeVariable).SomeField.context.Set`. The left part of the expression starting from the context is now a MemberAccess on a field, and not a Constant directly. I suspect that the more inner linq contexts are in the query, the more inner field there will be in the expression tree.

I wrote a correction locally, I can submit a pull request. The correction consists in replacing the code by a call to a method that retrieve the value of the expression recursively, going up in the tree until it finds a constant value

The original code is :

```
            var constant = expression as ConstantExpression;
            if (constant != null)
            {
                var value = constant.Value;
                if (value != null)
                {
                    return GetContextFromMember(member, value);
                }
            }
```

The new code could be :

```
           //Retrieve the context value from the encapsulated scope
            object value = GetExpressionValue(expression);
            if (value != null)
            {
                return GetContextFromMember(member, value);
            }
```

With the helper method as follow : 

```
        // <summary>
        // Tries to retrieve the value of an expression
        // If the expression is a constant, it returns the constant value.
        // If the expression is a field or property access on an expression, it returns the value of it recursively.
        // Otherwise it returns null
        // </summary>
        // <param name="expression">The expression</param>
        // <returns> The expression value. </returns>
        private static object GetExpressionValue(Expression expression)
        {
            //If the given expression is a constant, we just return its value
            ConstantExpression constantExpression = expression as ConstantExpression;
            if(constantExpression!=null)
            {
                return constantExpression.Value;
            }

            //If the given expression is a member access on an inner expression, we recursively reterieve the value of the inner expression, and get the member value from it.
            MemberExpression memberExpression = expression as MemberExpression;
            if(memberExpression!=null)
            {
                var asField = memberExpression.Member as FieldInfo;
                if (asField != null)
                {
                    object innerValue = GetExpressionValue(memberExpression.Expression);

                    if (innerValue != null)
                    {
                        return asField.GetValue(innerValue);
                    }
                }

                var asProperty = memberExpression.Member as PropertyInfo;
                if (asProperty != null)
                {
                    object innerValue = GetExpressionValue(memberExpression.Expression);

                    if (innerValue != null)
                    {
                        return asProperty.GetValue(innerValue, null);
                    }
                }
            }

            return null;
        }
```

**Comment by jcchalte** [comment]
anyone here ? @rowanmiller @maumar  ?

**Comment by jcchalte** [comment]
Hey ! Thank you for your answer ! Yes, a nightly build would unblock us, as our only issue is that we don't have your private key (obviously :) ) in order to generate a new package for our own purpose that would still work with 3rd party assemblies that also relies on EF DLLs.

**Comment by rowanmiller** [comment]
Our plan is to tackle most of the issues assigned to the 6.2 release in early December (just to give you an idea of timeline so that you aren't left hanging 😄). BTW also feel free to send a Pull Request if you want,  given you already did a lot of the work to isolate the issue.

**Comment by jcchalte** [comment]
Hello Rowan !

I've already sent a pull request ( #104 ) with the relevant modification. Technically, the correction should not have any side effect.

I am new to github and git, and did not find the way to create a pull request with only the DbQueryVisitor.cs modification. It also included an irrelevant .sln modification I had to do which is not related to the current bug. Should I create a new pull request with only the .cs or can you merge just the .cs file from the existing pull request ?

**Comment by rowanmiller** [comment]
Sorry I missed the PR you sent (and thank you for sending it!). We'll get someone to review it. The .sln changes look reasonable to me, but the developer that reviews your PR can confirm. We'll be processing all the outstanding PRs during the same timeframe I mentioned above.

**Comment by maumar** [comment]
Fixed in https://github.com/aspnet/EntityFramework6/commit/926744d447e3ddfa78a7f4af6c095373b7c8213a https://github.com/aspnet/EntityFramework6/commit/6463c603fe8b9f8132c7766c446fc76f58bf748e https://github.com/aspnet/EntityFramework6/commit/e7628ba43584da9d406d00726ecc6a96846f1209 https://github.com/aspnet/EntityFramework6/commit/f47edd8ba3e34515e8204090e18b9c1165e6fe71
